### DISCUSSION TITLE
FionaCDEServiceEngine/sidecar 404

### DISCUSSION THREAD

**Discussion by Mbucari** [mkb79/Audible#113] [Q&A] [answered]
I'm trying to use the `https://cde-ta-g7g.amazon.com/FionaCDEServiceEngine/sidecar` endpoint to retrieve bookmarks and clips, but I'm getting 404 errors. I started using by using AudibleApi project, [similar to how this method works](https://github.com/rmcrackan/AudibleApi/blob/b43fe2d41265fc38eb2a1949623192bb3a45e1d3/AudibleApi/Api.Activation.cs#L16-L32).

I used the following path: `/FionaCDEServiceEngine/sidecar?type=AUDI&key={asin}`, but I got a 404 response.

I then tried with audible-cli using this command:
`audible api -m GET -p type=AUDI -p key=B002UUKWCY https://cde-ta-g7g.amazon.com/FionaCDEServiceEngine/sidecar`
And I also got a 404.

Am I doing something wrong, or has Audible changed something?

**Comment by mkb79** [comment, marked as answer]
I‘ve tested it myself with `audible api -p type=AUDI -p key=B08M5FQ2CP https://cde-ta-g7g.amazon.com/FionaCDEServiceEngine/sidecar` and it will give me a 200 response with the data I requested.

**Reply by mkb79** [reply]
Oh wait. I'm think 404 in this case means there are no sidecar information available. Do you have heared this book yet?

**Reply by Mbucari** [reply]
I tried it on another title and it worked. Very strange!  Thank you for confirming it worked for you lol.

**Reply by Mbucari** [reply]
> Do you have heared this book yet?

I forgot to answer your question. I have listened to about 60% of B002UUKWCY, but I have not yet completed listening to it. I made 4 test bookmarks for that book, and I can see them all on my android device, but I am getting the 404 error.

I tried making one test bookmark on another book, which I have finished listening to, and I can retrieve that bookmark from the api.

EDIT:

I Just tested it with three more books, and it worked with all of them. So far, only B002UUKWCY has not worked.

**Reply by mkb79** [reply]
> Oh wait. I'm think 404 in this case means there are no sidecar information available. Do you have heared this book yet? 

I‘ve checked my notices and tried myself again. 404 means there are no annotations for the request. I've wrote about these [here](https://github.com/mkb79/Audible/issues/99#issuecomment-1176633221).

**Reply by mkb79** [reply]
> I tried making one test bookmark on another book, which I have finished listening to, and I can retrieve that bookmark from the api.

In the past I've changed the title of a clip. These changes where on my iOS device. But as I reinstalled the whole app these title are "gone". So I record the traffic to see what happens. As far as I could find out some annotations are not synced (e.g. title changes) and are local only.

**Reply by Mbucari** [reply]
>  As far as I could find out some annotations are not synced

That makes a lot of sense. Well actually it makes no sense from a design perspective, but it does explain what I'm seeing.  Thank you again for remeasuring me that I'm not making a mistake.

**Reply by mkb79** [reply]
> > As far as I could find out some annotations are not synced
> 
> That makes a lot of sense. Well actually it makes no sense from a design perspective, but it does explain what I'm seeing. Thank you again for remeasuring me that I'm not making a mistake.

You can try it out yourself and reinstall the app on your device (logout/login is not enough on iOS devices because the db is not deleted). Then you will see if the clip are synced. Or with less work install the app on another device (which can make use of clips) and you will see it too.

**Reply by Mbucari** [reply]
> You can try it out yourself 

I believe you. I'm just flabbergasted that Audible allowed this to happen.

**Comment by Mbucari** [comment]
@mkb79 It appears that this endpoint doesn't return clips. I only see "audible.bookmark",  "audible.note", and "audible.last_heard" types.

"audible.note" always returns an endPosition, but it is identical to "startPosition". Is there another endpoint or additional query parameters for getting clip details?

**Reply by mkb79** [reply]
You can set a clip with these body 

```
<annotations version="1.0" timestamp="2023-01-02T06:32:05+0100">
    <book key="B08M5FQ2CP" type="AUDI" version="23555421" guid="CR!JXYV7SZVQN381CD856CFJXHGHHPM:23555421" format="AAX_44_128"><last_heard action="modify" begin="7728497" timestamp="2023-01-02T06:32:01+0100" /><clip action="create" begin="7698497" end="7728497" timestamp="2023-01-02T06:32:05+0100"><metadata><![CDATA[{"c_version":"23555421"}]]></metadata></clip></book>
</annotations>
```

Here is a body to modify the clip

```

<annotations version="1.0" timestamp="2023-01-02T06:33:16+0100">
    <book key="B08M5FQ2CP" type="AUDI" version="23555421" guid="CR!JXYV7SZVQN381CD856CFJXHGHHPM:23555421" format="AAX_44_128"><clip action="delete" begin="7691167" end="7721167" timestamp="2022-12-21T07:02:56+0100"><metadata><![CDATA[{"c_version":"23555421"}]]></metadata></clip><clip action="create" begin="7659689" end="7779501" timestamp="2023-01-02T06:33:17+0100"><metadata><![CDATA[{"c_version":"23555421"}]]></metadata></clip></book>
</annotations>
```

And here is the result

```bash

marcel@nuc:~$ audible api -p type=AUDI -p key=B08M5FQ2CP https://cde-ta-g7g.amazon.com/FionaCDEServiceEngine/sidecar
{"md5": "XXX", "payload": {"records": [{"creationTime": "2023-01-02 05:33:18.0", "startPosition": "7659689", "lastModificationTime": "2023-01-02 05:33:18.0", "type": "audible.bookmark", "annotationId": "XXX"}, {"creationTime": "2023-01-02 05:33:17.0", "startPosition": "7659689", "lastModificationTime": "2023-01-02 05:33:17.0", "type": "audible.clip", "metadata": {"c_version": "23555421"}, "annotationId": "XXX", "endPosition": "7779501"}, {"creationTime": "2023-01-02 05:33:18.0", "startPosition": "7691167", "lastModificationTime": "2023-01-02 05:33:18.0", "type": "audible.bookmark", "annotationId": "XXX"}, {"creationTime": "2023-01-02 05:32:05.0", "startPosition": "7698497", "lastModificationTime": "2023-01-02 05:32:05.0", "type": "audible.bookmark", "annotationId": "XXX"}, {"creationTime": "2023-01-02 05:32:05.0", "startPosition": "7698497", "lastModificationTime": "2023-01-02 05:32:05.0", "type": "audible.clip", "metadata": {"c_version": "23555421"}, "annotationId": "XXX", "endPosition": "7728497"}, {"creationTime": "2023-01-02 05:32:28.0", "startPosition": "8364952", "lastModificationTime": "2023-01-02 05:32:28.0", "type": "audible.clip", "metadata": {"c_version": "23555421"}, "annotationId": "XXX", "endPosition": "8394953"}, {"creationTime": "2023-01-02 05:32:28.0", "startPosition": "8364952", "lastModificationTime": "2023-01-02 05:32:28.0", "type": "audible.bookmark", "annotationId": "XXX"}, {"creationTime": "2023-01-02 05:51:01.0", "startPosition": "9492485", "type": "audible.last_heard"}], "acr": null, "guid": "_LATEST_", "key": "B08M5FQ2CP", "type": "AUDI"}}
```

FYI:
Here is a body to modify the position last heared

```
<annotations version="1.0" timestamp="2023-01-02T06:32:25+0100">
    <book key="B08M5FQ2CP" type="AUDI" version="23555421" guid="CR!JXYV7SZVQN381CD856CFJXHGHHPM:23555421" format="AAX_44_128"><last_heard action="modify" begin="8392435" timestamp="2023-01-02T06:32:22+0100" /></book>
</annotations>
```

**Reply by mkb79** [reply]
Here are the body of a clip including a notice

```
<annotations version="1.0" timestamp="2023-01-02T07:54:17+0100">
    <book key="B08M5FQ2CP" type="AUDI" version="23555421" guid="CR!JXYV7SZVQN381CD856CFJXHGHHPM:23555421" format="AAX_44_128"><clip action="delete" begin="9462485" end="9492485" timestamp="2023-01-02T07:53:55+0100"><metadata><![CDATA[{"c_version":"23555421","note":"Test Clip notice"}]]></metadata></clip><clip action="create" begin="9462485" end="9540202" timestamp="2023-01-02T07:54:18+0100"><metadata><![CDATA[{"c_version":"23555421","note":"Test Clip notice"}]]></metadata></clip></book>
</annotations>
```

**Reply by Mbucari** [reply]
What is the version number "23555421"? Is it file-specific, or a constant?

**Reply by mkb79** [reply]
> What is the version number "23555421"? Is it file-specific, or a constant?

That’s the `version` key returned by a request to the `1.0/content/{asin}/metadata` endpoint.

**Reply by Mbucari** [reply]
I tried to create a clip. I got a 200 response, but when I queried the book I discovered that it had onle created a bookmark and not a clip.

Here is my POST body:

```XML
<annotations version="1.0" timestamp="2023-01-02T11:06:15-0700">
  <book key="B017WJ5ZK6" type="AUDI" version="10748071" guid="CR!FK7N7VFNKN03D6KYAD52P8H3CFNR:10748071" format="M4A_AAX">
    <last_heard action="modify" begin="1030000" timestamp="2023-01-02T11:06:15-0700" />
    <clip action="create" begin="1000000" end="1030000" timestamp="2023-01-02T11:06:15-0700">
      <metadata><![CDATA[{"c_version":"10748071"}]]></metadata>
    </clip>
  </book>
</annotations>
```

And here's what I get when I query that book on sidecar:

```JSON

{
   "md5":"f4d79e5f0b059272892926dec5726c21",
   "payload":{
      "type":"AUDI",
      "key":"B0983ZZZZM",
      "guid":"_LATEST_",
      "acr":null,
      "records":[
         {
            "annotationId":"aKYO4N975MVRM",
            "startPosition":"1000000",
            "type":"audible.bookmark",
            "creationTime":"2023-01-02 18:11:44.0",
            "lastModificationTime":"2023-01-02 18:11:44.0"
         },
         {
            "startPosition":"1030000",
            "type":"audible.last_heard",
            "creationTime":"2023-01-02 18:11:43.0"
         }
      ]
   }
}
```

Do you have any idea what's wrong?

**Reply by mkb79** [reply]
Can you try out to make a sidecar request with a guid version providing? In your case `CR!FK7N7VFNKN03D6KYAD52P8H3CFNR:10748071` or better `CR!FK7N7VFNKN03D6KYAD52P8H3CFNR`?

**Reply by Mbucari** [reply]
```JSON
{
   "md5":"XXX",
   "payload":{
      "guid":"CR!FK7N7VFNKN03D6KYAD52P8H3CFNR",
      "key":"B017WJ5ZK6",
      "type":"AUDI",
      "records":[
         {
            "startPosition":"1000000",
            "lastModificationTime":"2023-01-02 18:29:29.0",
            "creationTime":"2023-01-02 18:29:29.0",
            "type":"audible.bookmark",
            "annotationId":"XXX"
         },
         {
            "startPosition":"1030000",
            "creationTime":"2023-01-02 18:29:29.0",
            "type":"audible.last_heard"
         },
         {
            "startPosition":"1805583",
            "lastModificationTime":"2023-01-02 18:28:45.0",
            "creationTime":"2023-01-02 18:28:45.0",
            "type":"audible.bookmark",
            "annotationId":"XXX"
         }
      ],
      "acr":null
   }
}
```

Just bookmarks, no clips.

**Reply by Mbucari** [reply]
Interestingly, the clip that I set IS showing up inside the Audible app. It's just not being returned by sidecar.

**Reply by mkb79** [reply]
Maybe you are requested the wrong guid?!

**Reply by mkb79** [reply]
I‘ve tested it out myself and created a clip and the clip does appear in my app and my script.

**Reply by Mbucari** [reply]
I'm using the right guids. I've checked and rechecked. I even tried posting with guid="\_LATEST\_", and it worked (the clip appeared in my mobile app), but the clips are not returned by my query. Maybe it's an issue with my account?

I'm out of ideas.

**Reply by mkb79** [reply]
You can try it out with a new book which already have a clip set via app and try again. The book I used had a clip present before I add a new one via the api.

**Reply by Mbucari** [reply]
Good idea. I have a book with a clip that was added August 6, 2022. That clip is not returned by the query. Maybe a different endpoint is used for clips in US? I never managed to successfully man-in-the-middle the Audible app, so maybe I need to try that again and see exactly where it's getting the data from.

**Reply by mkb79** [reply]
I've an US account. Will try out tomorrow if there are any differences.

**Reply by mkb79** [reply]
One idea for today. The full url requested by the app is e.g. `https://cde-ta-g7g.amazon.com/FionaCDEServiceEngine/sidecar?format=AAX_44_128&guid=CR!JXYV7SZVQN381CD856CFJXHGHHPM:23555421&key=B08M5FQ2CP&type=AUDI`. Maybe you have to provide a format and guid query param. There are new formats beginning with `M4A_AAX_44_64`. But these are not used by the app. Maybe the format must be the same like in your post.

**Reply by Mbucari** [reply]
No luck. I tried with url `"/FionaCDEServiceEngine/sidecar?type=AUDI&key=B017WJ5ZK6&guid=CR!FK7N7VFNKN03D6KYAD52P8H3CFNR:10748071&format=M4A_AAX"`, and I still only get bookmarks and notes. I tried setting a clip with a notice like in your example. The note is returned by sidecar, but the clip isn't.

> There are new formats beginning with M4A_AAX_44_64

I'm using `M4A_AAX` as the format because it is what's returned in content_metadata. So far I have not seen any format strings that include bitrate information like `M4A_AAX_44_64`

**Comment by mkb79** [comment]
This is the response

```
{'content_metadata': {'content_reference': {'acr': 'CR!CZE5DS4AJ9577545Q0D779EGVGPN', 'asin': 'B08M5FQ2CP', 'codec': 'mp4a.40.2', 'content_format': 'M4A_AAX_44_64', 'content_size_in_bytes': 313872825, 'file_version': '1', 'marketplace': 'AN7V1F1VY261K', 'sku': 'BK_ADOP_301368', 'tempo': '1.0', 'version': '23555421'}}, 'response_groups': ['always-returned', 'content_reference']}
```

There seams to be a new codec and content_format naming scheme. 

Edit:
Here with the right acr 

```
{'content_metadata': {'content_reference': {'acr': 'CR!JXYV7SZVQN381CD856CFJXHGHHPM', 'asin': 'B08M5FQ2CP', 'codec': 'mp4a.40.2', 'content_format': 'AAX_44_128', 'content_size_in_bytes': 631947313, 'file_version': '2', 'marketplace': 'AN7V1F1VY261K', 'sku': 'BK_ADOP_301368', 'tempo': '1.0', 'version': '23555421'}}, 'response_groups': ['always-returned', 'content_reference']}
```

**Comment by Mbucari** [comment]
@mkb79 I want to thank you again for all of your help. I've been exploring this endpoint, and I think I've found the minimum amount of information required to create each of the four record types (last_heard, bookmark, note, and clip). Version appears to be optional, and "\_LATEST\_" is a valid guid.

This is a valid request:

```Xml
<annotations version="1.0" timestamp="2023-01-04T16:57:31-0700">
  <book key="B017WJ5ZK6" type="AUDI" guid="_LATEST_">
    <last_heard action="modify" begin="6000000" timestamp="2023-01-04T16:57:29-0700" />
    <clip action="create" begin="5000000" end="5030000" timestamp="2023-01-04T16:57:27-0700">
      <metadata><![CDATA[{"note":"Clip Note","title":"Clip Title"}]]></metadata>
    </clip>
    <note action="create" begin="6000000" end="6000000" timestamp="2023-01-04T16:57:28-0700">Standalone Note</note>
    <bookmark action="create" begin="7000000" timestamp="2023-01-04T16:57:28-0700" />
  </book>
</annotations>
```

And this is returned by the query:

```Json
{
    "md5": "XXX",
    "payload": {
        "type": "AUDI",
        "key": "B017WJ5ZK6",
        "acr": null,
        "records": [
            {
                "startPosition": "5000000",
                "creationTime": "2023-01-04 23:58:43.0",
                "annotationId": "XXX",
                "type": "audible.bookmark",
                "lastModificationTime": "2023-01-04 23:58:43.0"
            },
            {
                "startPosition": "5000000",
                "creationTime": "2023-01-04 23:57:55.0",
                "annotationId": "XXX",
                "type": "audible.clip",
                "endPosition": "5030000",
                "metadata": {
                    "note": "Clip Note",
                    "title": "Clip Title"
                },
                "lastModificationTime": "2023-01-04 23:57:55.0"
            },
            {
                "startPosition": "5000000",
                "creationTime": "2023-01-04 23:58:43.0",
                "annotationId": "XXX",
                "text": "Clip Note",
                "type": "audible.note",
                "endPosition": "5000000",
                "lastModificationTime": "2023-01-04 23:58:43.0"
            },
            {
                "startPosition": "6000000",
                "creationTime": "2023-01-04 23:57:56.0",
                "annotationId": "XXX",
                "text": "Standalone Note",
                "type": "audible.note",
                "endPosition": "6000000",
                "lastModificationTime": "2023-01-04 23:57:56.0"
            },
            {
                "startPosition": "6000000",
                "creationTime": "2023-01-04 23:57:57.0",
                "type": "audible.last_heard"
            },
            {
                "startPosition": "7000000",
                "creationTime": "2023-01-04 23:57:56.0",
                "annotationId": "XXX",
                "type": "audible.bookmark",
                "lastModificationTime": "2023-01-04 23:57:56.0"
            }
        ],
        "guid": "_LATEST_"
    }
}
```

Note: These two records are created automatically by Audible from the clip. Sometimes it took several seconds for them to be returned by the query after the clip was created, other times they were created instantly. The other 4 records were returned by the query immediately after creation.

```Json
           {
                "startPosition": "5000000",
                "creationTime": "2023-01-04 23:58:43.0",
                "annotationId": "XXX",
                "type": "audible.bookmark",
                "lastModificationTime": "2023-01-04 23:58:43.0"
            },
            {
                "startPosition": "5000000",
                "creationTime": "2023-01-04 23:58:43.0",
                "annotationId": "XXX",
                "text": "Clip Note",
                "type": "audible.note",
                "endPosition": "5000000",
                "lastModificationTime": "2023-01-04 23:58:43.0"
            },
```

**Reply by Mbucari** [reply]
Additionally, all 6 of those records may be deleted in one action using the following XML

```Xml
<annotations version="1.0" timestamp="2023-01-04T17:09:31-0700">
  <book key="B017WJ5ZK6" type="AUDI" guid="_LATEST_">
    <bookmark action="delete" begin="5000000" timestamp="2023-01-04T17:09:31-0700" />
    <clip action="delete" begin="5000000" timestamp="2023-01-04T17:09:31-0700" end="5030000" />
    <note action="delete" begin="5000000" timestamp="2023-01-04T17:09:31-0700" end="5000000" />
    <note action="delete" begin="6000000" timestamp="2023-01-04T17:09:31-0700" end="6000000" />
    <last_heard action="delete" begin="6000000" timestamp="2023-01-04T17:09:31-0700" />
    <bookmark action="delete" begin="7000000" timestamp="2023-01-04T17:09:31-0700" />
  </book>
</annotations>
```

EDIT* Actually, the bookmark and note that were created automatically from the clip remained after this delete action. I had to run a second delete on those 2 to remove them. I think their clip must be deleted before they can be deleted.

**Reply by mkb79** [reply]
I'm thank you for your work! This will hopefully enrich the use of Libation! If you need help in the future for other things feel free to contact me.
